main.go

```diff
@@ -106,22 +124,109 @@ func main() {
 			labelValue:             labelValue,
 		}
 		c := newController(klient, logger, opt)
+		c.registerMetrics(reg)
 		done := make(chan struct{})
 
 		g.Add(func() error {
 			return c.run(done)
-		}, func(err error) {
+		}, func(_ error) {
 			level.Info(logger).Log("msg", "shutting down controller")
 			close(done)
 		})
 	}
-	level.Info(logger).Log("msg", "starting the controller")
+	{
+		router := http.NewServeMux()
+		router.Handle("/metrics", promhttp.InstrumentMetricHandler(reg, promhttp.HandlerFor(reg, promhttp.HandlerOpts{})))
+		srv := &http.Server{Addr: config.InternalAddr, Handler: router}
 
+		g.Add(func() error {
+			return srv.ListenAndServe()
+		}, func(err error) {
+			if err != http.ErrServerClosed {
+				level.Warn(logger).Log("msg", "internal server closed unexpectedly")
+				return
+			}
+			level.Info(logger).Log("msg", "shutting down internal server")
+			ctx, _ := context.WithTimeout(context.Background(), internalServerShutdownTimeout)
+			if err := srv.Shutdown(ctx); err != nil {
+				stdlog.Fatal(err)
+			}
+		})
+	}
+
+	level.Info(logger).Log("msg", "starting the controller")
 	if err := g.Run(); err != nil {
 		stdlog.Fatal(err)
 	}
 }
 
+type prometheusReflectorMetrics struct {
```

Comment:
I would expect that there is an adapter like this somewhere in kubernetes

Reply:
Apperantly, there was one at a certain point (https://github.com/weiwei04/kubernetes/blob/fix_reflector_last_resource_version/pkg/util/reflector/prometheus/prometheus.go), but not anymore. Or I couldn't find where it lives right now

Reply:
I'm fairly certain kubernetes exposes these metrics, so we just need to find where it hides that implementation from us :)

Reply:
Very odd. I tried finding it as well, and really can't. We should ask around as this doesn't seem intentional.